### PR TITLE
bases: introduce abstract Base and buildd implementation

### PR DESCRIPTION
**Pull request by cjp256**
Depends on #13 and #14.

### CODE REVIEW

File: craft_providers/base.py [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+from abc import ABC, abstractmethod
+from typing import Optional
+
+from .executor import Executor
+
+logger = logging.getLogger(__name__)
+
+
+class Base(ABC):
+    """Base definition for bootstrap configuration and setup.
```
Comment (sergiusens) [inline craft_providers/base.py:26]: This docstring does not really tell me much

File: craft_providers/base.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    :param name: Name of base.
+    """
+
+    def __init__(self, *, name: str) -> None:
+        self.name = name
+
+    @abstractmethod
+    def setup(self, *, executor: Executor) -> None:
+        """Configure instance as necessary.
```
Comment (sergiusens) [inline craft_providers/base.py:36]:
```suggestion
        """Configure instance with the necessary environment.
```

But it begs the question, what is the necessary environment, or what should classes inheriting from this one implement

File: craft_providers/bases/craft_config.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+from craft_providers import Executor, errors
+
+from .errors import BaseConfigurationError
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass_json
+@dataclass
+class CraftBaseConfig:
```
Comment (sergiusens) [inline craft_providers/bases/craft_config.py:36]: can this module be its own PR?

File: craft_providers/base.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+import logging
+from abc import ABC, abstractmethod
+from typing import Optional
+
+from .executor import Executor
+
+logger = logging.getLogger(__name__)
+
+
+class Base(ABC):
```
Comment (sergiusens) [inline craft_providers/base.py:25]: can this be its own PR? It would have the benefit of being independent from the other ones too

**Review by sergiusens — COMMENTED** [review]
I would suggest leaving this PR as reference and splitting it up a bit more

**Comment by cjp256** [comment]
Breaking this up, i first started with Base @ https://github.com/canonical/craft-providers/pull/18.
